runtime/lua/vim/lsp/buf.lua

```diff
@@ -495,6 +495,69 @@ function M.outgoing_calls()
   call_hierarchy(ms.callHierarchy_outgoingCalls)
 end
 
+--- @param type_hierarchy_items lsp.TypeHierarchyItem[]?
+local function pick_type_hierarchy_item(type_hierarchy_items)
+  if not type_hierarchy_items then
+    return
+  end
+  if #type_hierarchy_items == 1 then
+    return type_hierarchy_items[1]
+  end
+  local items = {}
+  for i, item in pairs(type_hierarchy_items) do
+    local entry = item.detail or item.name
+    table.insert(items, string.format('%d. %s', i, entry))
+  end
+  local choice = vim.fn.inputlist(items)
+  if choice < 1 or choice > #items then
+    return
+  end
+  return choice
+end
+
+--- @param method string
+local function type_hierarchy(method)
+  local params = util.make_position_params()
+  --- @param result lsp.TypeHierarchyItem[]?
+  request(ms.textDocument_prepareTypeHierarchy, params, function(err, result, ctx)
```

Comment:
For proper multi-client handling this should use `buf_request_all`. The results from the servers can be merged/appended.

As it is right now, it could result in multiple prompts to select an item, and then the quickfix list would be populated once for each client/server.